### PR TITLE
Issue404 automatically validate process graph before download or execute

### PR DESCRIPTION
**Pull request by JohanKJSchreurs**
Implements https://github.com/Open-EO/openeo-python-client/issues/404

### CODE REVIEW

**Review by soxofaan — COMMENTED** [review]
some quick notes already

File: openeo/rest/connection.py [hunk 1]
```diff
@@ -1474,12 +1481,27 @@ def _build_request_with_process_graph(self, process_graph: Union[dict, FlatGraph
...
+                "Process graph is not valid. Validation errors:\n" + "\n".join(e["message"] for e in validation_errors)
+            )
+
     # TODO: unify `download` and `execute` better: e.g. `download` always writes to disk, `execute` returns result (raw or as JSON decoded dict)
     def download(
         self,
         graph: Union[dict, FlatGraphableMixin, str, Path],
         outputfile: Union[Path, str, None] = None,
         timeout: Optional[int] = None,
+        validate: Optional[bool] = True,
```
[the remark this answers is not in the document]
Comment (soxofaan) [reply openeo/rest/connection.py:1504]: unless there would be a connection level attribute to set a default `validate` value

[the remark this answers is not in the document]
Comment (JohanKJSchreurs) [reply openeo/rest/connection.py]: Removed Optional.

File: openeo/rest/connection.py [hunk 2]
```diff
@@ -1052,7 +1052,14 @@ def validate_process_graph(self, process_graph: dict) -> List[dict]:
...
         """
-        request = {"process_graph": process_graph}
+        # TODO: sometimes process_graph is already in the graph. Should we really *always* add it?
+        #   Was getting errors in some new unit tests because of the double process_graph but
+        #   perhaps the error is really not here but somewhere else that adds process_graph
+        #   when it should not? Still needs to be confirmed.
+        if "process_graph" not in process_graph:
+            request = {"process_graph": process_graph}
+        else:
+            request = process_graph
```
[the remark this answers is not in the document]
Comment (JohanKJSchreurs) [reply openeo/rest/connection.py]: Done.

File: openeo/rest/connection.py [hunk 3]
```diff
@@ -1474,12 +1481,27 @@ def _build_request_with_process_graph(self, process_graph: Union[dict, FlatGraph
         result["process"] = process_graph
         return result
 
+    def _warn_if_process_graph_invalid(self, process_graph: Union[dict, FlatGraphableMixin, str, Path]):
+        if not self.capabilities().supports_endpoint("/validation", "POST"):
+            return
+
+        graph = as_flat_graph(process_graph)
+        if "process_graph" not in graph:
+            graph = {"process_graph": graph}
```
[the remark this answers is not in the document]
Comment (JohanKJSchreurs) [reply openeo/rest/connection.py]: Done.

File: openeo/rest/connection.py [hunk 4]
```diff
@@ -1474,12 +1481,27 @@ def _build_request_with_process_graph(self, process_graph: Union[dict, FlatGraph
...
+
+        validation_errors = self.validate_process_graph(process_graph=graph)
+        if validation_errors:
+            _log.warning(
+                "Process graph is not valid. Validation errors:\n" + "\n".join(e["message"] for e in validation_errors)
+            )
```
[the remark this answers is not in the document]
Comment (JohanKJSchreurs) [reply openeo/rest/connection.py]: Done. 
Also added a test to verify that errors like this don't block the process, but that the expected warnings do appear in the logs.

**Comment by soxofaan** [comment]
merged in 78e8e939d77d63a9cd699c8861e648dda29c905b

thanks!

**Pull request closed without merging**